### PR TITLE
[FEATURE] Unexpected index list with/without values - New parameter in result_format

### PR DESCRIPTION
**Pull request by MMoussabbih**
Hello everyone,

This **PR** is related to the issue : https://github.com/great-expectations/great_expectations/issues/8780

It's about a new parameter `unexpected_metrics_with_values` _(True by default)_, in `result_format`, which allows us to choose, during the validation stage, whether we want an `unexpected_index_list` with column values or not and whether we want an `unexpected_list` or not _(an empty list)._

**_Use cases_** :

Let's consider that we are using `expect_column_values_to_match_regex` on `USER_EMAIL` column for **PII data** related to **customers**.

- **Use Case 1** : `unexpected_metrics_with_values = True`

```python
"result_format": { 
      "result_format": "COMPLETE",
      "unexpected_index_column_names": ["__id"],
      "unexpected_metrics_with_values": True,
}
``` 

We get : (the standard behaviour)

```python
"unexpected_index_list": [
      {
            "__id" : 14997,
            "USER_EMAIL": "john-doe@gmamil.com"
      },
      {
            "__id" : 15600,
            "USER_EMAIL": "johannadeo1988@outlook.cmo"
      },
      ...
],
"unexpected_list": [
      "john-doe@gmamil.com",
      "johannadeo1988@outlook.cmo",
      ...
]
```

- **Use Case 2** : `unexpected_metrics_with_values = False`

```python
"result_format": { 
      "result_format": "COMPLETE",
      "unexpected_index_column_names": ["__id"],
      "unexpected_metrics_with_values": False,
}
``` 

We get : _(the new feature)_

```python
"unexpected_index_list": [
      {
            "__id" : [14997, 15600, ...],
      }
],
"unexpected_list": [ ]
```

- **Use Case 3** : `unexpected_metrics_with_values` parameter is not mentioned

```python
"result_format": { 
      "result_format": "COMPLETE",
      "unexpected_index_column_names": ["__id"],
}
``` 

We get : _(the standard behaviour, since this new parameter is True by default)_

```python
"unexpected_index_list": [
      {
            "__id" : 14997,
            "USER_EMAIL": "john-doe@gmamil.com"
      },
      {
            "__id" : 15600,
            "USER_EMAIL": "johannadeo1988@outlook.cmo"
      },
      ...
],
"unexpected_list": [
      "john-doe@gmamil.com",
      "johannadeo1988@outlook.cmo",
      ...
]
```

**Checklist :** 

- [X] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [X] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [X] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [X] Appropriate tests and docs have been updated

### CODE REVIEW

**Comment by MMoussabbih** [comment]
@cla-bot check

File: docs/docusaurus/docs/reference/expectations/result_format.md
Comment (anthonyburdi) [inline docs/docusaurus/docs/reference/expectations/result_format.md]: Hi @MMoussabbih! Would you consider a change to the naming of the parameter to help clarify the functionality? Note that the new name also flips the default value.
```suggestion
- `exclude_unexpected_values`: When running validations, a set of unexpected results' indices and values is returned.  Setting this value to `True` suppresses values from the output to only have indices (default is `False`).
```

File: great_expectations/expectations/metrics/map_metric_provider/column_map_condition_auxilliary_methods.py
Comment (anthonyburdi) [inline great_expectations/expectations/metrics/map_metric_provider/column_map_condition_auxilliary_methods.py]: I'm not sure if this file is the place to implement this change, but would it be possible to suppress the `unexpected_list` in the output rather than an empty list? The concern is that an empty list would communicate that there are no unexpected values and could potentially be confusing.

So for example the result (when your new parameter is implemented) would be:
```
"unexpected_index_list": [
      {
            "__id" : [14997, 15600, ...],
      }
],
```
instead of:
```
"unexpected_index_list": [
      {
            "__id" : [14997, 15600, ...],
      }
],
"unexpected_list": [ ]
```

**Review by anthonyburdi — CHANGES_REQUESTED** [review]
Hi @MMoussabbih!! Thank you for spending the time to not only clearly document the issue you encountered but for also proposing a solution! I checked with a few others on the team and everyone is excited about your change. We only have a few minor changes requested before we can merge this PR.

File: docs/docusaurus/docs/reference/expectations/result_format.md
Comment (MMoussabbih) [reply docs/docusaurus/docs/reference/expectations/result_format.md]: Hello @anthonyburdi,
Thanks for your comment. I have updated the parameter name as you mentioned in my last commit 👍

File: great_expectations/expectations/metrics/map_metric_provider/column_map_condition_auxilliary_methods.py
Comment (MMoussabbih) [reply great_expectations/expectations/metrics/map_metric_provider/column_map_condition_auxilliary_methods.py]: I have made changes on `_format_map_output` function : In case of `exclude_unexpected_values=True`, the following attributes are suppressed : 

- `unexpected_list`
- `partial_unexpected_list`
- `partial_unexpected_counts` : _It contains values too..._

File: docs/docusaurus/docs/reference/expectations/result_format.md
Comment (Shinnnyshinshin) [inline docs/docusaurus/docs/reference/expectations/result_format.md:22]: Thank you for being thorough and including this change to the documentation. Much appreciated :)

File: tests/expectations/test_format_map_output.py
Comment (Shinnnyshinshin) [inline tests/expectations/test_format_map_output.py]: can we remove the `[NEW TEST]` comment before merging?

File: tests/expectations/metrics/column_map_metrics/test_unexpected_indices_and_query.py
Comment (Shinnnyshinshin) [inline tests/expectations/metrics/column_map_metrics/test_unexpected_indices_and_query.py:148]: Thank you for the marker here too

**Review by Shinnnyshinshin — APPROVED** [review]

**Review by Shinnnyshinshin — COMMENTED** [review]
Looks good to me. One ask before merging : could you remove the `[NEW TEST]` comments before merging?

**Review by anthonyburdi — APPROVED** [review]
Thank you for making the suggested changes! Just one more small item from @Shinnnyshinshin (removing the test comments) and we can merge. Approving now to remove my "Requested Changes" status.

File: tests/expectations/test_format_map_output.py
Comment (MMoussabbih) [reply tests/expectations/test_format_map_output.py]: Hello @Shinnnyshinshin ! Done 👍

**Comment by MMoussabbih** [comment]
Hello @anthonyburdi, @Shinnnyshinshin,
Thanks a lot for your reviews, I have deleted the `[NEW TEST]` comments.

**Comment by anthonyburdi** [comment]
Hi @MMoussabbih - we are nearly there - just seeing a few static-analysis (type checking) errors in CI tests: https://github.com/great-expectations/great_expectations/actions/runs/6497867075/job/17647907436?pr=8785

**Comment by MMoussabbih** [comment]
Hello @anthonyburdi,
I've made the necessary changes to resolve the static-analysis problems in CI tests.

**Pull request merged**
